spec.bs

```diff
@@ -510,16 +510,16 @@ A [=constructor string parser=] has an associated <dfn export for="constructor s
   <p>Second, the URLPattern constructor string parser needs to avoid applying URL canonicalization to all code points like [=basic URL parser=] does.  Instead we perform canonicalization on only parts of the pattern string we know are safe later when compiling each component pattern string.
   <p>Finally, the URLPattern constructor string parser does not handle some parts of the [=basic URL parser=] state machine.  For example, it does not treat backslashes specially as they would all be treated as pattern characters and would require excessive escaping.  In addition, this parser might not handle some more esoteric parts of the URL parsing algorithm like file URLs with a hostname.  The goal with this parser was to handle the most common URLs while allowing any niche case to be handled instead via the {{URLPatternInit}} constructor.
+  <p>In the constructor string algorithm, the pathname, search and hash are wildcarded if earlier components are specified but later ones are not. For example, "`https://example.com/foo`" matches any search and any hash. Similarly, "`https://example.com`" matches any URL on that origin. The username and password components are always wildcard unless they are explicitly specified.
 </div>
 
 <div algorithm>
 To <dfn>parse a constructor string</dfn> given a string |input|:
 
   1. Let |parser| be a new [=constructor string parser=] whose [=constructor string parser/input=] is |input| and [=constructor string parser/token list=] is the result of running [=tokenize=] given |input| and "<a for="tokenize policy">`lenient`</a>".
     <div class=note>
-      <p>When constructing a pattern using a {{URLPatternInit}} like `new URLPattern({ pathname: 'foo' })` any missing components will be defaulted to wildcards.  In the constructor string case, however, all components are precisely defined as either empty string or a longer value.  This is due to there being no way to simply "leave out" a component when writing a URL.
-      <p>To implement this we initialize components in |parser|'s [=constructor string parser/result=] with empty string in advance.
-      <p>We can't, however, do this immediately.  We want to allow the `baseURL` to provide information for relative URLs, so we only want to set the default empty string values for components following the first component in the relative URL.  We therefore wait to set the default component values until after we exit the "<a for="constructor string parser/state">`init`</a>" [=constructor string parser/state=].
+      <p>When constructing a pattern using a {{URLPatternInit}} like `new URLPattern({ pathname: 'foo' })` any missing components will be defaulted to wildcards.
+      <p>Components which appear "later" than those specified are instead treated as wildcards (see above). As a result, these values are not initialized to be empty in |parser|'s [=constructor string parser/result=] until a "later" component is seen.
```

Comment:
The new version of this note doesn't make a lot of sense. Previously, it was contrasting `new URLPattern({ pathname: 'foo' })` and the constructor string parser. Now, it's stating two disconnected facts: one about the `{ pathname: 'foo' }` version, and one about the constructor string parser. And for some reason it's using "instead" to contrast them, even though the two disconnected facts are actually very similar.

I think maybe this note could be deleted, now that the treatment is more symmetric? Or maybe something like the following:

> When constructing a pattern using a {{URLPatternInit}} like `new URLPattern({ pathname: 'foo' })` any missing components will be defaulted to wildcards. For the constructor string case, a somewhat-similar behavior applies: missing components are treated as either the empty string, or as wildcards, depending on what components _do_ appear in the input string. In particular, missing components which appear "later" than present components are treated as wildcards, whereas missing components that appear "earlier" than present components are treated as the empty string.
>
> To accomplish this behavior, missing components aren't set in the parser's result until a later component is encountered, as part of the change state algorithm.